README.md
## Thinking Analytics Unity SDK 使用指南

Thinking Analytics Unity SDK 提供了事件和用户属性数据采集相关接口，可以很方便的集成到您的项目中，向 TA 系统上报用户数据。

详细的使用指南请参考 Thinking Analytics 官网的[Unity SDK 使用指南](https://doc.thinkingdata.cn/tdamanual/installation/unity_sdk_installation.html)。

#### 1. 初始化 SDK
1.1 下载 [Unity SDK ](http://download.thinkingdata.cn/client/release/ThinkingAnalytics.unitypackage)资源文件，并导入资源文件到您的项目中：Assets > Import Package > Custom Package，选中您刚刚下载的文件

1.2 添加 ThinkingAnalytics GameObject, 并配置服务器地址和 APP ID

<img src="https://doc.thinkingdata.cn/tdamanual/assets/unity_sdk_installation.png" width = "50%"/>

>    注意：Android 插件使用 Gradle 集成，因此目前只支持 Unity 5.4 之后的版本。 

1.3 使用 SDK

当您配置好服务器地址和 APP ID 后，就可以开始使用 ThinkingAnalytics namespace 上传事件了，我们也提供了 Sample 供您参考。

```c#
using ThinkingAnalytics;

ThinkingAnalyticsAPI.Track("unity_start");
```

#### 2 设置用户 ID
在使用TDA的SDK之后，SDK会使用UUID作为每个用户的默认访客ID，该ID将会作为用户在未登录状态下身份识别ID。需要注意的是，UUID在用户重新安装APP以及更换设备时将会变更。

2.1 设置访客 ID（可选）

如果您的 APP 对每个用户有自己的访客ID管理体系，则您可以调用`Identify`来设置访客 ID:
```c#
ThinkingAnalyticsAPI.Identify("unity_id");
```

如果需要获得访客 ID，可以调用`GetDistinctId`获取：
```c#
ThinkingAnalyticsAPI.GetDistinctId();
```

2.2 设置账号 ID

在用户进行登录时，可调用`Login`来设置用户的账号 ID，在设置完账号 ID 后，将会以账号 ID 作为身份识别 ID，并且设置的账号 ID 将会在调用`Logout`之前一直保留：
```c#
// 设置账号 ID
ThinkingAnalyticsAPI.Login("unity_user");

// 清除账号 ID
ThinkingAnalyticsAPI.Logout();
```
> 注意：该方法不会上传用户登录、用户登出等事件。

#### 3 上传事件
通过`ThinkingAnalyticsAPI.Track()` 可以记录事件及其属性。一般情况下，您可能需要上传十几到上百个不同的事件，如果您是第一次使用 TDA 后台，我们推荐您先上传几个关键事件。

3.1 上传事件

建议您根据先前梳理的文档来设置事件的属性以及发送信息的条件。事件名称是`string`类型，只能以字母开头，可包含数字，字母和下划线“_”，长度最大为50个字符，对字母大小写不敏感。
- 事件属性是`Dictionary<string, object>` 类型，其中每个元素代表一个属性；
- 事件属性`Key`为属性名称，为`string`类型，规定只能以字母开头，包含数字，字母和下划线“_”，长度最大为50个字符，对字母大小写不敏感；
- 属性值支持四种类型：字符串、数值类、bool、DateTime。

```c#
Dictionary<string, object> properties = new Dictionary<string, object>()
    {
        {"KEY_DateTime", DateTime.Now.AddDays(1)},
        {"KEY_STRING", "B1"},
        {"KEY_BOOL", true},
        {"KEY_NUMBER", 50.65}
    };
ThinkingAnalyticsAPI.Track("TEST_EVENT", properties);
```

当您调用 `Track()` 时，SDK 会取系统当前时间作为 `#event_time` 属性值，如果您需要指定事件时间，可以传入 `DateTime`:
```c#
DateTime dateTime = DateTime.Now.AddDays(-1);
ThinkingAnalyticsAPI.Track("TEST_EVENT", properties, dateTime);
```

> 注意：尽管事件可以设置触发时间，但是接收端会做如下的限制：只接收相对服务器时间在前 10 天至后 3 天的数据，超过时限的数据将会被视为异常数据，整条数据无法入库。

v1.3.0+ 版本开始，SDK 支持上传事件的时间偏移（对应预置属性 #zone_offset），但是如果传入的 dateTime 的 Kind 属性为 DataTimeKind.Unspecified，则不会上报时间偏移。

3.2 设置公共属性

对于一些重要的属性，譬如用户的会员等级、来源渠道等，这些属性需要设置在每个事件中，此时您可以将这些属性设置为公共事件属性。公共事件属性指的就是每个事件都会带有的属性，您可以调用`SetSuperProperties`来设置公共事件属性，我们推荐您在发送事件前，先设置公共事件属性。

```c#
Dictionary<string, object> superProperties = new Dictionary<string, object>()
    {
        {"SUPER_LEVEL", 0},
        {"SUPER_CHANNEL", "A3"}
    };
ThinkingAnalyticsAPI.SetSuperProperties(superProperties);
```
公共事件属性将会被保存到缓存中，无需每次启动`APP`时调用。如果调用`SetSuperProperties`上传了先前已设置过的公共事件属性，则会覆盖之前的属性。如果公共事件属性和`Track()`上传的某个属性的`Key`重复，则该事件的属性会覆盖公共事件属性。

如果您需要删除某个公共事件属性，您可以调用`UnsetSuperProperty()`清除其中一个公共事件属性；如果您想要清空所有公共事件属性，则可以调用`ClearSuperProperties()`.

```c#
// 清除属性名为 SUPER_CHANNEL 的公共属性
ThinkingAnalyticsAPI.UnsetSuperProperty("SUPER_CHANNEL");

// 清空所有公共属性
ThinkingAnalyticsAPI.ClearSuperProperties();
```

如果公共属性的值不是常量，您可以通过设置动态公共属性的方式实现。动态公共属性的优先级大于公共事件属性。设置动态公共属性需要实现 IDynamicSuperProperties 接口。方式如下：
```c#
using ThinkingAnalytics;

// 定义动态公共属性实现，此例为设置 UTC 时间的动态公共属性
public class DynamicProp : IDynamicSuperProperties
{
    public Dictionary<string, object> GetDynamicSuperProperties()
    {
        return new Dictionary<string, object>() {
            {"KEY_UTCTime", DateTime.UtcNow}
        };
    }
}

ThinkingAnalyticsAPI.SetDynamicSuperProperties(new DynamicProp());
```

3.3 记录事件时长

您可以调用`TimeEvent()`来开始计时，配置您想要计时的事件名称，当您上传该事件时，将会自动在您的事件属性中加入`#duration`这一属性来表示记录的时长，单位为秒。
```c#
// 调用 TimeEvent 开启对 TIME_EVENT 事件的计时
ThinkingAnalyticsAPI.TimeEvent("TIME_EVENT");

// do some thing...

// 通过 Track 上传 TIME_EVENT 事件时，会在属性中添加 #duration 属性
ThinkingAnalyticsAPI.Track("TIME_EVENT");
```

#### 4 用户属性
TDA 平台目前支持的用户属性设置接口为`UserSet`、`UserSetOnce`、`UserAdd`、`UserDelete`。

4.1 UserSet

对于一般的用户属性，您可以调用`UserSet`来进行设置，使用该接口上传的属性将会覆盖原有的属性值，如果之前不存在该用户属性，则会新建该用户属性。
```c#
ThinkingAnalyticsAPI.UserSet(new Dictionary<string, object>()
    {
        {"USER_PROP_NUM", 0},
        {"USER_PROP_STRING", "A3"}
    });
```
与事件属性类似：
- 用户属性是`Dictionary<string, object>` 类型，其中每个元素代表一个属性；
- 用户属性`Key`为属性名称，为`string`类型，规定只能以字母开头，包含数字，字母和下划线“_”，长度最大为50个字符，对字母大小写不敏感；
- 用户属性值支持四种类型：字符串、数值类、bool、DateTime。

4.2 UserSetOnce

如果您要上传的用户属性只要设置一次，则可以调用`UserSetOnce`来进行设置，当该属性之前已经有值的时候，将会忽略这条信息：
```c#
ThinkingAnalyticsAPI.UserSetOnce(new Dictionary<string, object>()
    {
        {"USER_PROP_NUM", -50},
        {"USER_PROP_STRING", "A3"}
    });
```
> 注意：`UserSetOnce`设置的用户属性类型及限制条件与`UserSet`一致。

4.3 UserAdd

当您要上传数值型的属性时，您可以调用`UserAdd`来对该属性进行累加操作，如果该属性还未被设置，则会赋值`0`后再进行计算，可传入负值，等同于相减操作。
```c#
ThinkingAnalyticsAPI.UserAdd(new Dictionary<string, object>()
    {
        {"USER_PROP_NUM", -100.9},
        {"USER_PROP_NUM2", 10.0}
    });
```
> 注意： `UserAdd` 置的属性类型以及`Key`值的限制与`UserSet`一致，但`Value`只允许数值类型。

4.4 UserDelete

如果您要删除某个用户，可以调用`UserDelete`将这名用户删除，您将无法再查询该名用户的用户属性，但该用户产生的事件仍然可以被查询到。
```c#
ThinkingAnalyticsAPI.UserDelete();
```

4.5 UserUnset

如果您需要重置用户的某个属性，可以调用`UserUnset`将已经设置的属性删除。此接口支持传入字符串或者列表类型的参数:
```c#
// 删除单个用户属性
ThinkingAnalyticsAPI.UserUnset("userPropertyName");

// 删除多个用户属性
List<string> listProps = new List<string>();
listProps.Add("aaa");
listProps.Add("bbb");
listProps.Add("ccc");

ThinkingAnalyticsAPI.UserUnset(listProps);
```

CHANGELOG.md
**v2.5.0** 2022/10/20
- 优化代码

**v2.4.1** 2022/08/08
- 支持 Package Manager 接入
- 新增自动采集场景加载、卸载事件
- 优化代码

**v2.4.0** 2022/06/24
- 新增支持去重追加用户属性
- 新增支持自动采集事件回调
- 新增支持 iOS / Android 数据加密上报
- 优化代码

**v2.3.1** 2022/04/21
- 新增预置属性禁用开关
- 优化代码

**v2.3.0** 2022/02/28
- 全平台支持缓存事件、批量发送
- 优化代码

**v2.2.5** 2022/01/17
- 新增预置属性`#app_version`
- 支持 Switch 等平台，放开对支持平台的限制
- 简化初始化流程，放开 Unity 预置体限制
- 优化代码

**v2.2.4** 2021/11/15
- 支持复杂数据类型传输
- 自动采集事件支持自定义属性
- 优化代码

**v2.2.3** 2021/09/24
- 适配 WebGL 平台
- 优化代码

**v2.2.2** 2021/09/15
- 支持手动初始化
- 支持事件黑名单
- 优化代码

**v2.2.1** 2021/07/21
- 优化代码

**v2.2.0** 2021/06/16
- 适配低版本 Xcode 版本
- 适配 .NET 3.5 / 2017.x 环境
- 新增 C# 异常采集
- 调整事件属性、公共属性、预置属性的覆盖顺序
- 更新 Android SDK (v2.7.0)、iOS SDK (v2.7.0)

**v2.1.5** 2021/04/19
- 兼容低版本 .NET
- 代码优化


**v2.1.4** 2021/03/15
- 适配 iOS 14
- 适配 Android 11
- 新增预置属性`#bundle_id` (应用唯一标识)
- 优化网络信号获取逻辑


**v2.1.3** 2021/01/04
- 支持 UnityEditor 中上报数据
- 支持 PC 端游戏 ( Windows, MacOS )
- 多实例场景,支持代码设置默认实例


**v.2.1.2**(2020-10-31)
- autodata中信息放入track事件的properties
- 移除iOS bundle资源文件
- install,start事件以后立即flush
- 新增iOS 5G信息采集;更新5G情况的数据上报判断逻辑
- 调整未知网络数据上报逻辑调整
- 默认上报策略调整为所有网络情况均可上报
- 大量安卓日志打印OOM异常优化


**v2.1.1** (2020-08-25)
- 修复特殊事件不设置timeZone导致上报错误的#zone_offset的问题.

**v2.1.0** (2020-08-24)
- 支持首次事件, 允许传入自定义的 ID 校验是否首次上报
- 支持可更新、可重写的事件
- 优化#lib/#lib_version 字段为 Unity SDK 信息
- 升级原生 SDK 版本为 v2.6.0

**v2.0.8** (2020-06-28)
- 更新原生 Android SDK，解决极端情况下的空指针异常

**v2.0.7** (2020-06-23)
- 新增预置属性 #system_language
- 更新原生 SDK 版本号: Android v2.5.5, iOS v2.5.5
- Android 不再通过修改 gradle 配置来引入ThinkingAnalyticsSDK 依赖

**v2.0.6** (2020-06-17)
- 解决 iOS 32 位机型 long 型溢出的问题

**v2.0.5** (2020-05-19)
- 解决 2018.04 版本无法使用 ntp 校准时间的问题

**v2.0.4** (2020-05-15)
- 更新原生 iOS SDK 到 v2.5.2，解决上个版本中 Debug 模式问题

**v2.0.3** (2020-05-14)
- 更新原生 SDK 版本，适配 TA 后台埋点管理功能
- 优化代码，避免出现异常（可能会导致与其他崩溃采集 SDK 冲突）

**v2.0.2** (2020-04-17)
- 修复 Double 类型在自定义 CultureInfo 的场景下格式错误

**v2.0.1** (2020-04-14)
- 修复 Android 平台 DEBUG 模式上报事件的 BUG.

**v2.0.0** (2020-04-03)
- 修改自动采集逻辑，支持四种类型的自动采集事件：启动、关闭、崩溃、安装
- 支持 SDK 时间校准功能，支持时间戳校准和 NTP 服务器校准
- 允许为用户属性相关设置接口传入指定时间

**v1.4.3** (2020-03-19)
- 支持对数据 #time 属性的默认时区设置
- 更新原生 SDK 版本

**v1.4.2** (2020-02-21)
- 更新iOS 插件，修复之前版本bug

**v1.4.1** (2020-02-14)
- 适配 Unity 2019.3.1f1

**v1.4.0** (2020-02-11)
- 属性值支持 List / Array 类型
- 新增 UserAppend 接口
- 支持 Debug 模式数据校验
- 支持为每个实例单独配置接收端地址
- 去除本地数据格式校验

**v1.3.1** (2019-12-25)
- 修复 Android 4.3 以下版本退出超时
- 修复 未包含 iOS 开发环境时的报错

**v1.3.0** (2019-10-18)
- 新增 UserUnset 接口：用于重置用户属性.
- 事件预置属性新增时间偏移，适配多时区需求

**v1.2.2** (2019-10-11)
- 优化 iOS 平台上报逻辑，解决缓存数据库异常导致的重复上报

**v1.2.1** (2019-09-27)
- 解决老版本 .NET 的兼容性问题

**v1.2.0** (2019-08-31)
- 支持关闭/打开数据上报
- 支持暂停/继续数据上报
- 支持轻量级实例
- 修复2019.2.1f 版本设置网络类型失败
- 修复timeEvent 不准确问题
- Android/iOS SDK 升级为2.1.0

**v1.1.0** (2019-08-09)
- 支持获取设备ID
- 支持动态公共属性
- 支持安装事件采集
- Android SDK 默认升级为2.0.2
- iOS SDK 升级为2.0.1
- 支持上报缓存选项，用户可以选择主动调用StartTrack() 去开始上报

**v1.0.0** (2019-06-20)
- 支持访客ID 和用户账号的设置
- 支持事件和用户属性上报
- 支持 ta_app_start 和 ta_app_end 事件的自动上报
- 支持公共属性接口
- 支持 timeEvent 接口
- 支持多项目上报


Sample/TAExample.cs
using UnityEngine;
using UnityEngine.SceneManagement;
using ThinkingAnalytics;
using System.Collections.Generic;
using System;
using System.Threading;

public class TAExample : MonoBehaviour, IDynamicSuperProperties, IAutoTrackEventCallback
{


    public GUISkin skin;
    private Vector2 scrollPosition = Vector2.zero;
    //private static Color MainColor = new Color(0, 0,0);
    private static Color MainColor = new Color(84f / 255, 116f / 255, 241f / 255);
    private static Color TextColor = new Color(153f / 255, 153f / 255, 153f / 255);
    static int Margin = 20;
    static int Height = 60;
    static float ContainerWidth = Screen.width - 2 * Margin;
    // 动态公共属性接口
    public Dictionary<string, object> GetDynamicSuperProperties()
    {
        return new Dictionary<string, object>() 
        {
            {"DynamicProperty", DateTime.Now}
        };
    }
    // 自动采集事件回调接口
    public Dictionary<string, object> AutoTrackEventCallback(int type, Dictionary<string, object>properties)
    {
        return new Dictionary<string, object>() 
        {
            {"AutoTrackEventProperty", DateTime.Today}
        };
    }

    private void Awake()
    {
    }
    private void Start()
    {
        // 自动初始化 ThinkingAnalytics 时，可以在 Start() 中调用 EnableAutoTrack() 开启自动采集事件
        // 开启自动采集事件
        // ThinkingAnalyticsAPI.EnableAutoTrack(AUTO_TRACK_EVENTS.ALL);
    }

    void OnGUI() 
    {
        GUILayout.BeginArea(new Rect(Margin, Screen.height * 0.15f, Screen.width-2*Margin, Screen.height));
        scrollPosition = GUILayout.BeginScrollView(new Vector2(0, 0), GUILayout.Width(Screen.width - 2 * Margin), GUILayout.Height(Screen.height - 100));
        GUIStyle style = GUI.skin.label;
        style.fontSize = 25;
        GUILayout.Label("Initialization / UserIDSetting",style);

        GUIStyle buttonStyle = GUI.skin.button;
        buttonStyle.fontSize = 20;
        GUILayout.BeginHorizontal(GUI.skin.box,GUILayout.Height(Height));
        if (GUILayout.Button("ManualInitialization", GUILayout.Height(Height)))
        {
            // 1. 手动初始化（已加载 ThinkingAnalytics 预置体）
            ThinkingAnalyticsAPI.StartThinkingAnalytics();


            // 2. 手动初始化（动态挂载 ThinkingAnalyticsAPI 脚本）
            // new GameObject("ThinkingAnalytics", typeof(ThinkingAnalyticsAPI));

            // 2.1 设置实例参数
            // string appId = "22e445595b0f42bd8c5fe35bc44b88d6";
            // string serverUrl = "https://receiver-ta-dev.thinkingdata.cn";
            // ThinkingAnalyticsAPI.StartThinkingAnalytics(appId, serverUrl);

            // 2.1 个性化设置实例参数
            // string appId = "22e445595b0f42bd8c5fe35bc44b88d6";
            // string serverUrl = "https://receiver-ta-dev.thinkingdata.cn";
            // ThinkingAnalyticsAPI.TAMode mode = ThinkingAnalyticsAPI.TAMode.NORMAL;
            // ThinkingAnalyticsAPI.TATimeZone timeZone = ThinkingAnalyticsAPI.TATimeZone.Local;
            // ThinkingAnalyticsAPI.Token token = new ThinkingAnalyticsAPI.Token(appId, serverUrl, mode, timeZone);
            // 开启加密传输（仅支持iOS/Android）
            // token.enableEncrypt = true;
            // token.encryptVersion = 0;
            // token.encryptPublicKey = "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCIPi6aHymT1jdETRci6f1ck535n13IX3p9XNLFu5xncfzNFl6kFVMiMSXMIwWSW2lF6ELtIlDJ0B00qE9C02n6YbIAV+VvVkchydbWrm8VdnEJk/6tIydoUxGyM9pDT6U/PaoEiItl/BawDj3/+KW6U7AejYPij9uTQ4H3bQqj1wIDAQAB";
            // ThinkingAnalyticsAPI.StartThinkingAnalytics(token);

            // 2.2 多项目支持
            // string appId_2 = "1b1c1fef65e3482bad5c9d0e6a823356";
            // string serverUrl_2 = "https://receiver-ta-dev.thinkingdata.cn";
            // ThinkingAnalyticsAPI.TAMode mode_2 = ThinkingAnalyticsAPI.TAMode.NORMAL;
            // ThinkingAnalyticsAPI.TATimeZone timeZone_2 = ThinkingAnalyticsAPI.TATimeZone.Local;
            // ThinkingAnalyticsAPI.Token token_2 = new ThinkingAnalyticsAPI.Token(appId_2, serverUrl_2, mode_2, timeZone_2);

            // ThinkingAnalyticsAPI.Token[] tokens = new ThinkingAnalyticsAPI.Token[2];
            // tokens[0] = token;
            // tokens[1] = token_2;
            // ThinkingAnalyticsAPI.StartThinkingAnalytics(tokens);

            // 多项目发送事件
            // ThinkingAnalyticsAPI.Track("test_event");
            // ThinkingAnalyticsAPI.Track("test_event_2", appId:appId_2);


            // 开启自动采集事件
            //ThinkingAnalyticsAPI.EnableAutoTrack(AUTO_TRACK_EVENTS.ALL);
            // 开启自动采集事件，并设置自定属性
            // ThinkingAnalyticsAPI.EnableAutoTrack(AUTO_TRACK_EVENTS.ALL, new Dictionary<string, object>() 
            // {
            //     {"auto_track_key", "auto_track_value"}
            // });
            // 开启自动采集，并设置事件回调
            // ThinkingAnalyticsAPI.EnableAutoTrack(AUTO_TRACK_EVENTS.ALL, this);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("SetAccountID", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.Login("TA");
        }
        GUILayout.Space(20);
        if (GUILayout.Button("SetDistinctID", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.Identify("TA_Distinct1");
        }
        GUILayout.Space(20);
        if (GUILayout.Button("ClearAccountID", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.Logout();
        }
        GUILayout.EndHorizontal();

        GUILayout.Space(20);
        GUILayout.Label("EventTracking", GUI.skin.label);
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("TrackEvent", GUILayout.Height(Height)))
        {
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties["channel"] = "ta";//字符串
            properties["age"] = 1;//数字
            properties["isVip"] = true;//布尔
            properties["birthday"] = DateTime.Now;//时间
            properties["object"] = new Dictionary<string, object>() { { "key", "value" } };//对象
            properties["object_arr"] = new List<object>() { new Dictionary<string, object>() { { "key", "value" } } };//对象组
            properties["arr"] = new List<object>() { "value" };//数组

            ThinkingAnalyticsAPI.Track("TA",properties);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("TrackFirstEvent", GUILayout.Height(Height)))
        {
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties["status"] = 1;
            TDFirstEvent firstEvent = new TDFirstEvent("DEVICE_FIRST", properties);
            ThinkingAnalyticsAPI.Track(firstEvent);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("TrackUpdateEvent", GUILayout.Height(Height)))
        {
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties["status"] = 2;
            TDUpdatableEvent updatableEvent = new TDUpdatableEvent("UPDATABLE_EVENT", properties, "test_event_id");
            ThinkingAnalyticsAPI.Track(updatableEvent);
        }

        GUILayout.Space(20);
        if (GUILayout.Button("TrackOverwriteEvent", GUILayout.Height(Height)))
        {
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties["status"] = 3;
            TDOverWritableEvent overWritableEvent = new TDOverWritableEvent("OVERWRITABLE_EVENT", properties, "test_event_id");
            ThinkingAnalyticsAPI.Track(overWritableEvent);
        }

        GUILayout.EndHorizontal();
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("TrackEventWithTimeTravel", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.TimeEvent("TATimeEvent");
            #if !(UNITY_WEBGL)
            Thread.Sleep(1000);
            #endif
            ThinkingAnalyticsAPI.Track("TATimeEvent");
        }

        GUILayout.Space(20);
        if (GUILayout.Button("TrackEventWithDate", GUILayout.Height(Height)))
        {
            Dictionary<string, object> properties = new Dictionary<string, object>();
            properties["status"] = 4;
            ThinkingAnalyticsAPI.Track("TA_001", properties, DateTime.Now.AddHours(-1));
        }

        GUILayout.Space(20);
        if (GUILayout.Button("TrackEventWithLightInstance", GUILayout.Height(Height)))
        {
            // 创建轻实例，返回轻实例的 token （类似于 APP ID）
            string lightToken = ThinkingAnalyticsAPI.CreateLightInstance();
            ThinkingAnalyticsAPI.Login("light_account", lightToken);
            ThinkingAnalyticsAPI.Track("light_event", lightToken);
        }
        GUILayout.EndHorizontal();




        GUILayout.Space(20);
        GUILayout.Label("UserProperty", GUI.skin.label);
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("UserSet", GUILayout.Height(Height)))
        {
            Dictionary<string, object> userProperties = new Dictionary<string, object>();
            userProperties["age"] = 1;
            ThinkingAnalyticsAPI.UserSet(userProperties);
        }

        GUILayout.Space(20);
        if (GUILayout.Button("UserSetOnce", GUILayout.Height(Height)))
        {
            Dictionary<string, object> userProperties = new Dictionary<string, object>();
            userProperties["gender"] = 1;
            ThinkingAnalyticsAPI.UserSetOnce(userProperties);

        }
        GUILayout.Space(20);
        if (GUILayout.Button("UserAdd", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.UserAdd("usercoin", 1);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("UserUnset", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.UserUnset("usercoin");
        }
        GUILayout.Space(20);
        if (GUILayout.Button("UserDelete", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.UserDelete();
        }
        GUILayout.Space(20);
        if (GUILayout.Button("UserAppend", GUILayout.Height(Height)))
        {
            List<string> propList = new List<string>();
            propList.Add("ball");
            Dictionary<string, object> userProperties = new Dictionary<string, object>();
            userProperties["prop"] = propList;
            ThinkingAnalyticsAPI.UserAppend(userProperties);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("UserUniqAppend", GUILayout.Height(Height)))
        {
            List<string> propList = new List<string>();
            propList.Add("apple");
            Dictionary<string, object> userProperties = new Dictionary<string, object>();
            userProperties["prop"] = propList;
            ThinkingAnalyticsAPI.UserUniqAppend(userProperties);
        }
        GUILayout.EndHorizontal();

   

        GUILayout.Space(20);
        GUILayout.Label("OtherSetting", GUI.skin.label);
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("Flush", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.Flush();
        }
        GUILayout.Space(20);
        if (GUILayout.Button("GetDeviceID", GUILayout.Height(Height)))
        {
            Debug.Log("DeviceID: " + ThinkingAnalyticsAPI.GetDeviceId());
        }
        GUILayout.Space(20);
        if (GUILayout.Button("Pause", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.SetTrackStatus(TA_TRACK_STATUS.PAUSE);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("Stop", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.SetTrackStatus(TA_TRACK_STATUS.STOP);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("SaveOnly", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.SetTrackStatus(TA_TRACK_STATUS.SAVE_ONLY);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("Normal", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.SetTrackStatus(TA_TRACK_STATUS.NORMAL);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("CalibrateTime", GUILayout.Height(Height)))
        {
            //时间戳,单位毫秒 对应时间为1608782412000 2020-12-24 12:00:12
            ThinkingAnalyticsAPI.CalibrateTime(1608782412000);

            //NTP 时间服务器校准，如：time.apple.com
            //ThinkingAnalyticsAPI.CalibrateTimeWithNtp("time.apple.com");
        }
        GUILayout.EndHorizontal();

        GUILayout.Space(20);
        GUILayout.Label("PropertiesSetting", GUI.skin.label);
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("SetSuperProperties", GUILayout.Height(Height)))
        {
            Dictionary<string, object> superProperties = new Dictionary<string, object>();
            superProperties["vipLevel"] = 1;
            ThinkingAnalyticsAPI.SetSuperProperties(superProperties);
        }
        GUILayout.Space(20);
        if (GUILayout.Button("UpdateSuperProperties", GUILayout.Height(Height)))
        {
            Dictionary<string, object> superProperties = new Dictionary<string, object>();
            superProperties["vipLevel"] = 2;
            ThinkingAnalyticsAPI.SetSuperProperties(superProperties);
        }
       
        GUILayout.Space(20);
        if (GUILayout.Button("ClearSuperProperties", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.UnsetSuperProperty("vipLevel");
        }

        GUILayout.Space(20);
        if (GUILayout.Button("ClearAllSuperProperties", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.ClearSuperProperties();
        }

        GUILayout.EndHorizontal();
        GUILayout.BeginHorizontal(GUI.skin.textArea, GUILayout.Height(Height));
        if (GUILayout.Button("SetDynamicSuperProperties", GUILayout.Height(Height)))
        {
            ThinkingAnalyticsAPI.SetDynamicSuperProperties(this);
        }

        GUILayout.Space(20);
        if (GUILayout.Button("GetPresetProperties", GUILayout.Height(Height)))
        {
            TDPresetProperties presetProperties = ThinkingAnalyticsAPI.GetPresetProperties();
            string deviceModel = presetProperties.DeviceModel;
            Debug.Log("TDPresetProperties DeviceModel is " + deviceModel);
            Dictionary<string, object> eventPresetProperties = presetProperties.ToEventPresetProperties();
            string propertiesStr = "";
            foreach (KeyValuePair<string, object> kv in eventPresetProperties)
            {
                propertiesStr = propertiesStr + kv.Key + " = " + kv.Value + ", ";
            }
            Debug.Log("eventPresetProperties: " + propertiesStr);
        }
        GUILayout.EndHorizontal();
        GUILayout.EndScrollView();
        GUILayout.EndArea();
    }
    
    private void OnEnable()
    {
        // 监听场景加载、卸载事件
        SceneManager.sceneLoaded += ThinkingAnalyticsAPI.OnSceneLoaded;
        SceneManager.sceneUnloaded += ThinkingAnalyticsAPI.OnSceneUnloaded;
    }
}
